pywinauto/unittests/test_application_linux.py

```diff
@@ -56,72 +58,75 @@ def test_start_problem(self):
-            app.connect(path=_test_app())
-            self.assertEqual(app.process, self.subprocess_app.pid)
-            app.kill()
+            self.app.connect(path=_test_app())
+            self.assertEqual(self.app.process, self.subprocess_app.pid)
 
         def test_get_cpu_usage(self):
-            app = Application()
-            app.start(_test_app())
+            self.app.start(_test_app())
             time.sleep(1)
-            self.assertGreater(app.cpu_usage(), 0)
-            app.kill()
+            self.assertGreater(self.app.cpu_usage(), 0)
 
         def test_is_process_running(self):
-            app = Application()
-            app.start(_test_app())
+            self.app.start(_test_app())
             time.sleep(1)
-            self.assertTrue(app.is_process_running())
-            app.kill()
-
-        def test_killed_app_not_running(self):
-            app = Application()
-            app.start(_test_app())
-            time.sleep(1)
-            app.kill()
-            self.assertFalse(app.is_process_running())
+            self.assertTrue(self.app.is_process_running())
+            self.app.kill()
+            self.assertFalse(self.app.is_process_running())
 
         def test_kill_killed_app(self):
-            app = Application()
-            app.start(_test_app())
+            self.app.start(_test_app())
             time.sleep(1)
-            app.kill()
-            self.assertTrue(app.kill())
+            self.app.kill()
+            self.assertTrue(self.app.kill())
 
         def test_kill_connected_app(self):
             self.subprocess_app = subprocess.Popen(_test_app().split(), stdout=subprocess.PIPE, shell=False)
             time.sleep(1)
-            app = Application()
-            app.connect(pid=self.subprocess_app.pid)
-            app.kill()
+            self.app.connect(pid=self.subprocess_app.pid)
+            self.app.kill()
 
             # Unlock the subprocess explicity, otherwise
             # it's presented in /proc as a zombie waiting for
             # the parent process to pickup the return code
             self.subprocess_app.communicate()
             self.subprocess_app = None
 
-            self.assertFalse(app.is_process_running())
+            self.assertFalse(self.app.is_process_running())
+
+
+    class WindowSpecificationTestCases(unittest.TestCase):
+
+        """Unit tests for the application.WindowSpecification class"""
+
+        def setUp(self):
+            """Set some data and ensure the application is in the state we want"""
+            self.app = Application()
+
+        def tearDown(self):
+            """Close the application after tests"""
+            self.app.kill()
+
+        def test_app_binding(self):
+            self.app.start(_test_app())
+            self.assertEqual(self.app.NonExistingDialog.app, self.app)
+            self.assertEqual(self.app.Application.Panel.exists(), True)
+            self.assertEqual(self.app.Application.Panel.app, self.app)
```

Comment:
A little surprise to me that it works this way. I mean access keywords in the window spec instance by attribute. OK. Just thought to run some real action or check for existence.

Another thought is that WindowSpecification must be created and saved to variable before starting new instance in the same `app` object. Because creating spec after re-start shouldn't catch the issue, should it?

Reply:
> A little surprise to me that it works this way. I mean access keywords in the window spec instance by attribute.

It was always the case that internal attributes are looked through before going to resolve any external attribute: https://github.com/pywinauto/pywinauto/blob/4005bff2aacb6e5421a43abdbdd2f4c178ca53b4/pywinauto/base_application.py#L365


> Another thought is that WindowSpecification must be created and saved to variable before starting new instance in the same app object.

Apparently, an old instance of WindowSpecification will continue running as far as the app instance is the same. If the app restarts different program the spec will fail to resolve the described control, but if we restart the same program the description is still valid and will be resolved even for a different PID. I think this behavior is just fine. But I'll add another test with Application restart.

Reply:
Yes, thanks. But I thought this another test is the core case.